### ISSUE TITLE
spec coverage of virtual dispatch and default values

### ISSUE THREAD

**Issue by cassella** [chapel-lang/chapel#17679]
> I think we need to update the spec with our decision here - identical signatures needs to include intent, type and existence of default value - but not the default value itself.

_Originally posted by @mppf in https://github.com/chapel-lang/chapel/issues/13962#issuecomment-589830421_

The spec does now mention intent and type, in classes.rst, but doesn't mention default valueness;

> In order to have identical signatures, two methods must have the same
names, and their formal arguments must have the same names, intents, types,
and order.

As a language user though, one of the ways of a method and its
overload differing wrt default valueness make intuitive sense.  And
that case does work as I'd expect on master today, when the parent
class method has no default but the child class method does:

```chapel
var five = 5;
var six = 6;

class C {
  proc foo(x: int) { writeln("C.foo: ", x); }
}

class D : C {
  override proc foo(x: int = six) { writeln("D.foo: ", x); }
}

var c = new C();
c.foo(1);

var d = new D();
d.foo();
d.foo(2);

var d2: C = new D();
d2.foo(3);
```

```
C.foo: 1
D.foo: 6
D.foo: 2
D.foo: 3
```

The other way around compiles and runs, but isn't as intuitive to me:

```chapel
var five = 5;
var six = 6;

class C {
  proc foo(x: int = five) { writeln("C.foo: ", x); }
}

class D : C {
  override proc foo(x: int) { writeln("D.foo: ", x); }
}

var c = new C();
c.foo();
c.foo(1);

var d = new D();
d.foo();
d.foo(2);

var d2: C = new D();
d2.foo();
d2.foo(3);
```

```
C.foo: 5
C.foo: 1
D.foo: 5
D.foo: 2
D.foo: 5
D.foo: 3
```

As a language user, I'm especially surprised that the call to
`d.foo()` compiles.  Through #7880 et. al. I think I've picked up that
the method call's default values are computed at the call site via
virtual dispatch to something like `c.foo_default_x_method()`.  The
behavior makes sense in light of that.

But I don't think that's an understanding that's granted by the
documentation.  Though I can guess now that `D` has inherited `C`'s
`foo_default_x_method()`, and the compiler seeing that existing
inserts a call to it for `d.foo()`?

In `procedures.rst`, where the spec covers default values, it doesn't
suggest that default values are provided by compiler-generated
functions/methods:

> If the actual argument is omitted from the function call, the default expression is evaluated when the function call is made and the evaluated result is passed to the formal argument as if it were passed from the call site.

If both of these cases should work this way, then the spec section on
signatures doesn't need an update for this issue.  But I think it
should say somewhere that an overriding function inherits default
values, if not spelling out the existence and semantics of default
value methods.

If one or both of these cases shouldn't compile, the spec section on
signatures and the compiler both need updating.

**Comment by mppf** [comment]
huh, that is odd. I would think that this second case -
``` chapel
class C {
  proc foo(x: int = five) { writeln("C.foo: ", x); }
}

class D : C {
  override proc foo(x: int) { writeln("D.foo: ", x); }
}
```

should fail to compile because `D.foo` is not overriding `C.foo` in the case that the default value is provided. It seems strange indeed to take the default value from the parent class method declaration and use it to call the child class method.

**Comment by bradcray** [comment]
I agree with @casella's intuition and @mppf's diagnosis here.  I like that the first case is supported and works; and I think that the override in the second shouldn't be permitted.

Another semi-related case that feels tempting to support (but doesn't work today) is one in which the child class adds additional arguments as the overridden parent method, but supplies defaults for them.  For example:

```chapel
class C {
  proc foo() { writeln("C.foo"); }
}

class D : C {
  override proc foo(x: int = 6) { writeln("D.foo: ", x); }
}

var c = new C();
c.foo();
//c.foo(1);                                                                     

var d = new D();
d.foo();
d.foo(2);

var d2: C = new D();
d2.foo();
//d2.foo(3);                                                                    
```

Is there a deep reason not to support this case?  Implementation-wise, it seems like the biggest challenge would be override checking.  Though maybe this also causes more problems in the new default arguments world than it would've had in the old one?

**Comment by casella** [comment]
I'm afraid @cassella <> @casella 😄

**Comment by mppf** [comment]
> Is there a deep reason not to support this case? Implementation-wise, it seems like the biggest challenge would be override checking. Though maybe this also causes more problems in the new default arguments world than it would've had in the old one?

it is similar to #17683 in that when doing the virtual dispatch we need to be able to call a function with 0 arguments which means that we'd have to create a sortof wrapper function to be the function pointer to call with 0 arguments that fills in the defaults and then calls the real function.

**Comment by bradcray** [comment]
Aargh, I'm once again sorry @casella (and @cassella).  You'd think my github auto-complete would keep me more on-track here.
